chore: Remove error suppression for cppcheck

mdeweerd opened the pull request:
Remove suppressions of specific notices by cppcheck.

#4494 was updated to ignore these warnings to allow #4494 to be integrated in the master branch.

mdeweerd committed:
Qual: Enable ignored notices

mdeweerd commented:
FYI, the notices that are remaining:

```
src/uncrustify.cpp(1862): warning (nullPointerRedundantCheck): Either the condition 'ref=ref->GetPrev()' is redundant or there is possible null pointer dereference: ref.
src/keywords.cpp(199): style (unsignedLessThanZero): Checking if unsigned expression 'len' is less than zero.
src/keywords.cpp(217): error (uninitvar): Uninitialized variables: &key.type, &key.lang_flags
src/tokenizer/combine_fix_mark.cpp(1420): style (knownConditionTrueFalse): Condition 'tmp2->IsNotNullChunk()' is always false
src/align/var_def_brace.cpp(129): style (knownConditionTrueFalse): Condition 'pc->IsPreproc()' is always false
src/tokenizer/combine_labels.cpp(285): style (knownConditionTrueFalse): Condition 'tmp->IsNullChunk()' is always false
src/unc_text.cpp(43): style (constParameterReference): Parameter 'c0' can be declared as reference to const
src/align/var_def_brace.cpp(129): style (knownConditionTrueFalse): Condition 'pc->IsPreproc()' is always false
src/tokenizer/mark_question_colon.cpp(147): style (knownConditionTrueFalse): Condition 'pc2->IsNotNullChunk()' is always false
src/tokenizer/mark_question_colon.cpp(102): style (knownConditionTrueFalse): Condition 'colon_found' is always false
```

guy-maurel commented:
I will provide the necessary changes

mdeweerd committed:
Make sure the cache directory exists for cppcheck executino (in pre-commit)

mdeweerd committed:
:wrench: Update cppcheck configuration with UNCRUSTIFY_VERSION flag

Added UNCRUSTIFY_VERSION flag to cppcheck command to avoir static analysis notice

gmaurel reviewed (changes requested):
Do not make any change at the file [.pre-commit-config.yaml 
I am working on it, I need about 10 days until I am ready.
Thanks

mdeweerd commented:
> Do not make any change at the file [.pre-commit-config.yaml I am working on it, I need about 10 days until I am ready. Thanks

This PR is mainly to verify rhe results of cppcjeck.
If you can add the extra option for cppcjeck that is useful.

mdeweerd commented:
Result of local execution:

```
src/tokenizer/mark_question_colon.cpp(156): style (knownConditionTrueFalse): Condition 'colon_found' is always false
src/uncrustify_emscripten.cpp(59): information (missingInclude): Include file: "uncrustify_version.h" not found.
src/uncrustify_emscripten.cpp(176): performance (passedByValueCallback): Function parameter 'tag' should be passed by const reference. However it seems that '_add_keyword' is a callback function.
src/uncrustify_emscripten.cpp(226): performance (passedByValueCallback): Function parameter 'name' should be passed by const reference. However it seems that 'reset_option' is a callback function.
src/uncrustify_emscripten.cpp(252): performance (passedByValueCallback): Function parameter 'name' should be passed by const reference. However it seems that 'set_option' is a callback function.
src/uncrustify_emscripten.cpp(252): performance (passedByValueCallback): Function parameter 'value' should be passed by const reference. However it seems that 'set_option' is a callback function.
src/uncrustify_emscripten.cpp(294): performance (passedByValueCallback): Function parameter 'name' should be passed by const reference. However it seems that 'get_option' is a callback function.
src/uncrustify_emscripten.cpp(116): style (constParameterReference): Parameter 'o' can be declared as reference to const
```

mdeweerd committed:
:wrench: Update cppcheck version to 2.20.1

Update the cppcheck version from 2.18.3 to 2.20.1 in the pre-commit workflow.

mdeweerd committed:
:wrench: Update Cppcheck cache key to include version

Reusing cppcheck cache from a different version seems to result in some trouble.
So the cache key was updated to include the Cppcheck version (${CPPCHECK_VERSION}).

mdeweerd committed:
:wrench: Update pre-commit workflow to clean cache before restore

mdeweerd committed:
:wrench: Update pre-commit hook versions

Updated beautysh from v6.4.1 to v6.4.3 and codespell from v2.4.1 to v2.4.2 to ensure compatibility with the latest features and bug fixes.

mdeweerd committed:
Qual: Exclude .cache directory explicitly

Excluding the .cache directory explicitly from pre-commit
analysis - in theory not needed, but checking if this
avoids the internal analysis errors for cppcheck

mdeweerd committed:
Qual: Update cppcheck version

mdeweerd committed:
Ignore notices about missing initializers on member variables

mdeweerd commented:
FYI: I've submitted a [PR](https://github.com/pocc/pre-commit-hooks/pull/73) to fix the cppcheck caching issue with the pre-commit hook.
A temporary solution could be to point to my fork for this fix.

gmaurel commented:
I have finished all the necessary changes.
The test with the file .pre-commit-config.yaml works pretty well for cppcheck.

May I close your call?

mdeweerd commented:
@gmaurel I've updated the description of this  PR, and squashed my commits.

I am waiting for a [PR for the cppcheck hook fixing caching](https://github.com/pocc/pre-commit-hooks/pull/73) to be accepted in that hook's release.  We could temporarily point the hook to my fork of pre-commit-hooks.
This will allow proper caching and increase the speed of the analysis - the "official" version for the cppcheck hook analyses the files one by one, rather than through a single cppcheck call and breaks any caching opportunity.
-> Do we wait, or point to my fork (when integrated we can point back to the upstream)?

The other updates here are the upgrade of some of the other pre-commit hook versions and the update of cppcheck to the latest version which brings some new cases:
And that it enables cppcheck for the big files - which will not be an issue when caching actually works.  If this is to be integrated before, I'll remove it again.

```
src/token_is_within_trailing_return.cpp(16): style (constParameterPointer): Parameter 'pc' can be declared as pointer to const
src/unc_tools.cpp(63): style (constParameterPointer): Parameter 'pc_sub' can be declared as pointer to const
src/unc_tools.cpp(833): style (constVariablePointer): Variable 'pc_saved' can be declared as pointer to const
src/detect.cpp(155): style (constVariablePointer): Variable 'pc' can be declared as pointer to const
```

Messages like the follwoing are related to the caching limitation of the current cppcheck hook: `nofile(0): error (internalError): failed to load '.cache/cppcheck/enum.a1' from '.cache/cppcheck/files.txt'`
